### ISSUE TITLE
Incorrect offsets?

### ISSUE THREAD

**Issue by Ben3874** [djkaty/Il2CppInspector#25]
Been working on generating a script when I noticed some of the offsets I was getting were incorrect. Code just loops through every type, gets a C++ equivalent type name then a field to its parent.
I've been using the Dummy Dll's from il2cppdumper to check offsets, and I noticed the offsets for the GUID class are incorrect.
What my code is outputting:
```
AddMember "Guid", "int32_t", "_a", 0, 0x10
AddMember "Guid", "int16_t", "_b", 0, 0x14
AddMember "Guid", "int16_t", "_c", 0, 0x16
AddMember "Guid", "uint8_t", "_d", 0, 0x18
AddMember "Guid", "uint8_t", "_e", 0, 0x19
AddMember "Guid", "uint8_t", "_f", 0, 0x1A
AddMember "Guid", "uint8_t", "_g", 0, 0x1B
AddMember "Guid", "uint8_t", "_h", 0, 0x1C
AddMember "Guid", "uint8_t", "_i", 0, 0x1D
AddMember "Guid", "uint8_t", "_j", 0, 0x1E
AddMember "Guid", "uint8_t", "_k", 0, 0x1F
```
The expected offsets
```
		[FieldOffset(Offset = "0x0")]
		private int _a;

		[FieldOffset(Offset = "0x4")]
		private short _b;

		[FieldOffset(Offset = "0x6")]
		private short _c;

		[FieldOffset(Offset = "0x8")]
		private byte _d;

		[FieldOffset(Offset = "0x9")]
		private byte _e;

		[FieldOffset(Offset = "0xA")]
		private byte _f;

		[FieldOffset(Offset = "0xB")]
		private byte _g;

		[FieldOffset(Offset = "0xC")]
		private byte _h;

		[FieldOffset(Offset = "0xD")]
		private byte _i;

		[FieldOffset(Offset = "0xE")]
		private byte _j;

		[FieldOffset(Offset = "0xF")]
		private byte _k;
```
Later on this breaks x64dbg, since sizeof(GUID) = 0x20, which makes the following impossible.
```
AddMember "Guid+GuidResult", "Guid", "parsedGuid", 0, 0x10
AddMember "Guid+GuidResult", "int32_t", "throwStyle", 0, 0x20
```
Note, that Guid.GuidResult.Guid  is also at 0x10, when it should be 0x0
More classes seem to be offset by 0x10 than those that are correct.

Download link: (Too big to upload to github) [https://drive.google.com/file/d/12L4pupCjxPStTm4AD3L-t8NwJAhkBsOm/view?usp=sharing](url)

**Comment by djkaty** [comment]
Okay, I spent a while looking at this, I believe the field offsets are correct. There was a recent commit to Il2CppDumper which changed the way field offsets are calculated:

```cs
                if (offset > 0)
                {
                    if (isValueType)
                    {
                        if (Is32Bit)
                        {
                            offset -= 8;
                        }
                        else
                        {
                            offset -= 16;
                        }
                    }
                }
```

where `isValueType` refers to the type that the field is defined in (not the field's type).

Type instances in IL2CPP are represented by an `Il2CppObject` (typedef'd as a `RuntimeObject`):

```cpp
typedef Il2CppClass Il2CppVTable;
typedef struct Il2CppObject
{
    union
    {
        Il2CppClass *klass;
        Il2CppVTable *vtable;
    };
    MonitorData *monitor;
} Il2CppObject;
```
(il2cpp-object-internals.h)

This is followed by each field for the instance. Therefore a Guid instance will begin with 2 pointers (8 bytes each in 64-bit binaries), plus the 16 bytes of fields, for a total of 32 bytes (or 24 in 32-bit binaries, with the first field offset being 0x8 instead of 0x10).

You can see this in action in Field.cpp:

```cpp

    void Field::GetValue(Il2CppObject *obj, FieldInfo *field, void *value)
    {
        void *src;

        IL2CPP_ASSERT(obj);

        IL2CPP_ASSERT(!(field->type->attrs & FIELD_ATTRIBUTE_STATIC));

        src = (char*)obj + field->offset;
        SetValueRaw(field->type, value, src, true);
    }
````

However, when fields that are value types are being written, something different happens:

(from Field::SetValueRaw)
```cpp
            case IL2CPP_TYPE_VALUETYPE:
                /* note that 't' and 'type->type' can be different */
                if (type->type == IL2CPP_TYPE_VALUETYPE && Type::IsEnum(type))
                {
                    t = Class::GetEnumBaseType(Type::GetClass(type))->type;
                    goto handle_enum;
                }
                else
                {
                    Il2CppClass *klass = Class::FromIl2CppType(type);
                    int size = Class::GetValueSize(klass, NULL);
                    if (value == NULL)
                    {
                        memset(dest, 0, size);
                    }
                    else
                    {
                        memcpy(dest, value, size);
                        gc::GarbageCollector::SetWriteBarrier(reinterpret_cast<void**>(dest), size);
                    }
                }
                return;
```

The number of bytes to write at the offset is determined by Class::GetValueSize:

```cpp
    int32_t Class::GetValueSize(Il2CppClass *klass, uint32_t *align)
    {
        int32_t size;

        if (!klass->init_pending)
            Init(klass);

        IL2CPP_ASSERT(klass->valuetype);

        size = Class::GetInstanceSize(klass) - sizeof(Il2CppObject);

        if (align)
            *align = klass->minimumAlignment;

        return size;
    }
```
(from Class.cpp)

Notice, it is calculated as the full size of the instance, minus the size of the header of 2 pointers. Therefore, it would seem that a standalone instance of Guid is 8/16 bytes of pointers + 16 bytes of field data, but when it appears as a field, the pointers are elided.

I haven't investigated with Il2CPP behaves like that and It's absolutely possible I overlooked or misunderstood something, as I just delved into this now over my morning coffee and it's quite the rabbit hole, but thank you for putting me through this learning exercise 😂

In a nutshell, the size of value type fields is 8 bytes (32-bit) or 16 bytes (64-bit) less than their instance size, and as far as I can see the change to Il2CppDumper makes the offsets "correct" when the value type appears as a field, and wrong in all other cases.

**Comment by Perfare** [comment]
ValueType does not inherit from Il2CppObject in il2cpp, so FieldOffset of ValueType is always 
``` = field->offset - sizeof(Il2CppObject)```
In addition, ```fieldOffsets``` in il2cpp are not the final values used. You can see that its fieldoffset is always 0 for a generic class. Because for generic classes, it is not known whether ```T``` is a value type or a reference type. So il2cpp will recalculate fieldOffsets of all classes at runtime.

**Comment by djkaty** [comment]
Ahh I see, thanks. Why do they always store the field offsets starting at 0x8 or 0x10 if there is no non-field data in value types though?

Side note the compiler can know whether the generic type parameters are value or reference types if the programmer specifies type constraints for them, but of course that's not enforced in .NET.

**Comment by Perfare** [comment]
I think this may just be the early code, it has not been deleted, it has become redundant data.
There was another reason I forgot to say. In most environments, `sizeof(int) = 4`, but in other environments, value of `sizeof(int)` may be 2 or other, so it is necessary to recalculate the field offset at runtime. In this case, it does not matter what the field offset stored in il2cpp was before

**Comment by djkaty** [comment]
Okay, I've made the suggested changes so the offsets should be correct now :)

@Perfare you need to change your field offset calculation to exclude static fields, they're stored with the class rather than the instance (obviously) and always start at zero, so the static fields in your value types have the wrong offsets at the moment (I think, you can correct me again :P)

**Comment by Perfare** [comment]
@djkaty Okay, thank you for your reminder.
